### ISSUE TITLE
Identifier 'React' has already been declared with multiple external libraries

### ISSUE THREAD

**Issue by derpoho** [remix-run/remix#2987]
### What version of Remix are you using?

1.4.1

### Steps to Reproduce

More Details:
React Version: 18
Yarn

- Install remix for cloudflare pages setup
- Add libraries like (react-spring, framer-motion, react-hook/mouse-postion)
- Use them (simple import is enough)

### Expected Behavior

Application/Website loads correctly when running `wrangler pages dev ./public` and `remix watch`

### Actual Behavior

Website loads, shows for a few ms, then breaks. Console sends multiple errors:

```
Uncaught SyntaxError: Identifier 'React' has already been declared
```

```
Uncaught Error: Cannot initialize 'routeModules'. This normally occurs when you have server code in your client modules.
```

```
Hydration failed because the initial UI does not match what was rendered on the server.
```

```
react-dom.development.js:86 Warning: An error occurred during hydration. The server HTML was replaced with client content in <#document>.
```

The first error shows the error into the installed packages from above, always the same line (the first react import)
`var React = __toESM(require_react());`

This happens to a lot of modules. You can fix that by renaming React to something else within the package and patching those with npx patch-package. But this makes things hard as you have to patch a lot of packages. It looks like something is going wrong during compilation (which naturally is part of remix - non-configurable).

Can somebody reproduce this issue?

Edit:
Node Version: v16.13.2
Mac M1 Max

**Comment by kiliman** [comment]
I've had this issue as well. I believe it's an issue with `esbuild` bundler. If you minify the code, it seems to work.

https://github.com/evanw/esbuild/issues/475#issuecomment-946760991

**Comment by nickjs** [comment]
I can also reproduce this issue with `threejs` and `@react-three/drei`. I can prepare a very simple repro repo if helpful.

**Comment by agcty** [comment]
I can also also reproduce this issue with threejs and react-three-fiber as well.

**Comment by agcty** [comment]
It seems like esbuild is the issue in this case (https://github.com/evanw/esbuild/issues/475). Is there an option for remix to pass config options to esbuild without needing to patch packages? @kiliman

**Comment by jaschaio** [comment]
Here is an incredible ugly workaround, based on a [comment in the esbuiold issue](https://github.com/evanw/esbuild/issues/475#issuecomment-748528799):

[Modify your package.json](https://stackoverflow.com/a/70971915/2485910):

```diff
  "scripts": {
-   "build": "remix build",
+   "build": "NODE_OPTIONS='-r ./esbuild-overwrite' remix build",
-   "dev": "remix dev",
+   "dev": "NODE_OPTIONS='-r ./esbuild-overwrite' remix dev"
  },
```

Create a file `esbuild-overwrite.js` with the Following contents:

```js
const esbuild = require( 'esbuild' );
const Module = require( 'module' );

const originalRequire = Module.prototype.require;
const originalBuild = esbuild.build;

const reactShim = [
    ...Object.keys( require( 'react' ) ).map( k => `export const ${ k } = window.React.${ k }` ),
    'const React = window.React; export default React;',
].join( '\n' );
const reactDomShim = Object.keys( require( 'react-dom' ) ).map( k => `export const ${ k } = window.ReactDOM.${ k }` ).join( '\n' );

// https://esbuild.github.io/plugins/#using-plugins
const externaliseReactPlugin = {
    name: 'react',
    setup( build ) {
        
        build.onResolve( { filter: /^react$/ }, ( args ) => ( {
            path: args.path,
            namespace: 'react-ns',
        } ) );

        build.onLoad( { filter: /.*/, namespace: 'react-ns' }, ( args ) => ( {
            contents: reactShim,
            loader: 'js',
        } ) );
        
        build.onResolve( { filter: /^react-dom$/ }, ( args ) => ( {
            path: args.path,
            namespace: 'react-dom-ns',
        } ) );

        build.onLoad( { filter: /.*/, namespace: 'react-dom-ns' }, ( args ) => ( {
            contents: reactDomShim,
            loader: 'js',
        } ) );

    }
};

function build( options ) {

    if ( options.platform !== 'browser' )
        return originalBuild( options );

    return originalBuild( {
        ...options,
        external: [
            ...( options.external ) ? options.external : [],
            'react',
            'react-dom',
        ],
        plugins: [
            ...( options.plugins ) ? options.plugins : [],
            externaliseReactPlugin,
        ],
    } );

};

Module.prototype.require = function ( id ) {

    // when remix requires esbuild, it will get our modified build function from above
    if ( id === 'esbuild' )
        return { ...esbuild, build };

    return originalRequire.apply( this, arguments );

};
```

Add `react` and `react-dom` as global objects to your `window`, for example via script tags within your `root.jsx` file:

```diff
export default App() {

    return (return (
        <html lang="en">
            <head>
                <meta charSet="utf-8" />
                <meta name="viewport" content="width=device-width,initial-scale=1" />
                <Meta />
                <Links />
+               <script crossorigin src="https://unpkg.com/react@18/umd/react.production.min.js"></script>
+               <script crossorigin src="https://unpkg.com/react-dom@18/umd/react-dom.production.min.js"></script>/head>
            <body>
                <Outlet />
                <ScrollRestoration />
                <Scripts />
                <LiveReload />
            </body>
        </html>
    );

}
```

**Comment by kiliman** [comment]
Interesting. I think we need to determine if this is indeed an esbuild issue or not, and if so, what can be done to fix the root cause.

But yeah, anything that unblocks you sounds good to me.

**Comment by agcty** [comment]
> Interesting. I think we need to determine if this is indeed an esbuild issue or not, and if so, what can be done to fix the root cause.
> 
> But yeah, anything that unblocks you sounds good to me.

I agree, we definitively need to be able to say this is either a remix, esbuild or library issue.

Been trying to bundle libraries in a different way and see when this issue occurs, will let you know when I've got something more solid.

**Comment by kiliman** [comment]
Has anyone tried using `esbuild` with these libraries outside of Remix to see if the issue occurs? That would be the true test. If it works there, then it may be how Remix is splitting client/server bundles that's breaking this.

**Comment by jaschaio** [comment]
In my case I am not using any third party library but just an internal component library (that has it's own `package.json`). If I don't use remix and just esbuild it works. I only get the problem when using remix + esbuild

**Comment by kiliman** [comment]
Awesome. Thanks for the deep dive. I'll take a look when I get a chance. I know this is an annoying issue for many developers.

**Comment by MichaelDeBoey** [comment]
This problem is probably fixed by #3588

**Comment by jacob-ebey** [comment]
> This problem is probably fixed by #3588 & can be tested by [`v1.6.2-pre.0`](https://github.com/remix-run/remix/releases/tag/v1.6.2-pre.0)

This doesn't fix it unfortunately. That was the `minifyIdentifiers` flag, not the `minifySyntax` in that PR.

**Comment by agcty** [comment]
> > This problem is probably fixed by #3588 & can be tested by [`v1.6.2-pre.0`](https://github.com/remix-run/remix/releases/tag/v1.6.2-pre.0)
> 
> This doesn't fix it unfortunately. That was the `minifyIdentifiers` flag, not the `minifySyntax` in that PR.

So does this mean a fix isn't possible for now or this is a "won't fix" type of issue?

> Edit: I remembered wrong on the react-three-fiber issue, that is the minifyIdentifiers flag that fixes the duplicate React identifiers. minifyIdentifiers isn't something we want to enable in dev right now. (https://github.com/remix-run/remix/pull/3588)

**Comment by chaance** [comment]
> So does this mean a fix isn't possible for now or this is a "won't fix" type of issue?

Definitely looking to fix this. It's still at the top of my priority list.

**Comment by jsellam** [comment]
Thank you, I have the same issue with web3 https://www.npmjs.com/package/web3 on client side (used in useCallback).
I have this message : Error: Cannot initialize 'routeModules'. This normally occurs when you have server code in your client modules.

**Comment by timothymalcham** [comment]
I'm also experiencing this same error (Identifier 'React' has already... and Cannot initialize 'routeModules') when trying to use https://www.npmjs.com/package/@grapecity/activereports-react on the client side only

**Comment by joshuaellis** [comment]
hey @chaance,

Not sure how much you've looked into it, so I'm sorry if i'm gonna say what you already know. For context I maintain two libraries `react-spring` and `react-three-fiber` and set up the bundling for them, what's interesting is `react-spring` works perfectly fine with `remix` so that sort of scratched my itch to why one library with parallel bundling (using preconstruct) would not work.

I've had a look and got `@react-three/fiber` working with `remix` now by removing `@babel/runtime` from it's dependencies. I can also confirm that these other libraries – `@react-three/drei` & `web3` both have `@babel/runtime` in their dependencies most likely confirming this is the problem.

As for weather this is a `remix` or `esbuild` issue, i'll leave that up to you knowledge! But I hope this helps you narrow down the problem 👌🏼 

@ me if you have any questions // need any help etc.

**Comment by chaance** [comment]
This won't make it in time for the next release, but initial testing shows the fix should land when https://github.com/remix-run/remix/pull/3860 is merged.

**Comment by mcansh** [comment]
> This won't make it in time for the next release, but initial testing shows the fix should land when #3860 is merged.

if anyone here wants to take `0.0.0-experimental-ea38b6458` for a spin, that'd be great!

**Comment by lswest** [comment]
> > This won't make it in time for the next release, but initial testing shows the fix should land when #3860 is merged.
> 
> if anyone here wants to take `0.0.0-experimental-ea38b6458` for a spin, that'd be great!

I ran into this issue today with two packages' internal dependencies: `use-onclickoutside` (internal dependency `use-latest`), and `@reach/dialog` (internal dependency `react-focus-lock`). Somehow these two packages were yielding an additional React definition. 

### Results
Installing the above version allowed the pages to work properly again!

### Extra information

*The below is just extra information for anyone wanting to debug/investigate this further, you can safely ignore this otherwise.*

I figured out the packages by following the console.log error regarding duplicate definitions (one linked to the root bundle, one to another chunk). Then based on the comment in the file with the path to the file, I figured out which packages they were and which dependencies required them via `npm list`. Leaving this information here for anyone who needs it.

Side note regarding the issue appearing: The odd thing is I didn't touch either of those dependencies since the last time the code was working (last Friday, the 29th of July) and today. In remix 1.4.2 only one page wasn't working (related to `@reach/dialog`, upon updating to 1.6.7 no pages were working (the issue appeared in `use-onclickoutside`, which I use in the site header's popover navigation). Presumably the esbuild version was updated between those versions?

**Comment by mcansh** [comment]
before the update coming in #3860, esbuild has been at `0.14.22` since february 17th (<https://github.com/remix-run/remix/commit/50e428b65d0ec1ae72a19ea8a06bd8012fb01ce5>) which was done in [1.2.2](https://github.com/remix-run/remix/compare/v1.2.1...v1.2.2)

**Comment by lswest** [comment]
Thanks for the info. No idea why the problem waited until today to appear then or what else I might have done to trigger it, but the main thing is the experimental build fixes it 😄

**Comment by jaschaio** [comment]
> > This won't make it in time for the next release, but initial testing shows the fix should land when #3860 is merged.
> 
> if anyone here wants to take `0.0.0-experimental-ea38b6458` for a spin, that'd be great!

Looks like it works, thanks a lot!

Any idea which release this will hit?

**Comment by kokotom** [comment]
I have just tried 0.0.0-experimental-ea38b6458 and still not working for me :(

**Comment by mcansh** [comment]
@kokotom same error message? any particular dependencies you've narrowed down?

**Comment by lswest** [comment]
@kokotom - just to check (because I made this mistake the first time too) - did you change to the exact version from @mcansh's post? I had `^0.0.0-experimental-ea38b6458` in my package.json, so npm install grabbed a different version that it felt was "newer". Make sure it's just the version number (`0.0.0-experimental-ea38b6458`) in package.json.

**Comment by mcansh** [comment]
@lswest @kokotom yeah i just tried the latest experimental and all is fine with that library :) 

[demo](https://pnpm-turborepo-vercel-remix-f65bs34pf-mcansh.vercel.app)
[commit](https://github.com/mcansh/pnpm-turborepo-vercel-remix/commit/5847d74d0ffb447930674d03f104320fd9565d73)

**Comment by kokotom** [comment]
> @lswest @kokotom yeah i just tried the latest experimental and all is fine with that library :)
> 
> [demo](https://pnpm-turborepo-vercel-remix-f65bs34pf-mcansh.vercel.app) [commit](https://github.com/mcansh/pnpm-turborepo-vercel-remix/commit/5847d74d0ffb447930674d03f104320fd9565d73)

Oh man terribly sorry.
The package was:
@syncfusion/ej2-react-gantt

however I noticed that I didn't pin the version as mentioned by @lswest.
After doing so I'm getting a different error:

`Warning: React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.`

**Comment by saranshgrover** [comment]
`experimental-ea38b6458` solved it for me. Any ideas when this would be pushed onto a proper release?

**Comment by mcansh** [comment]
> `experimental-ea38b6458` solved it for me. Any ideas when this would be pushed onto a proper release?

finishing up some windows testing before i merge the PR just to be safe, but it should go out this week unless there's something i missed in the esbuild changelog that would prevent that.

**Comment by Swapnull** [comment]
> > @lswest @kokotom yeah i just tried the latest experimental and all is fine with that library :)
> > [demo](https://pnpm-turborepo-vercel-remix-f65bs34pf-mcansh.vercel.app) [commit](https://github.com/mcansh/pnpm-turborepo-vercel-remix/commit/5847d74d0ffb447930674d03f104320fd9565d73)
> 
> Oh man terribly sorry. The package was: @syncfusion/ej2-react-gantt Glad you noticed it yourself :)
> 
> However I noticed that I didn't pin the version as mentioned by @lswest. After doing so I'm getting a different error:
> 
> `Warning: React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.`
> 
> It happened both with react 17.0.2 and 18.2.0

I have just tried after having these same problems and get this same error. 
![image](https://user-images.githubusercontent.com/4456346/184081693-288d6b33-9caa-47c0-ad75-0750e59c37b1.png)
My entry.server.tsx is exactly the same as the ones from the examples. I am on react 17.0.2.

**Comment by raminrez** [comment]
In my case problem was importing the hole react module in one of the components
`import * as React from 'react'`

**Comment by patrikholcak** [comment]
@raminrez All the `import`s in a module should be isolated to just that module. It’s the compiler’s job to figure out the best way to rename/merge the import statements when concatenating the modules into one bundle.

What you’re seeing is a prime example of this bug. You shouldn’t be forced to rename a variable just because it occurs somewhere else in the bundle (perhaps even a library you didn’t write)

**Comment by chaance** [comment]
Looks like we were finally able to get this resolved! Updated to v1.7.0-pre.0 [here](https://github.com/chaance/remix-three-test) and everything runs swimmingly. We should be able to close this after the stable release is published 🥳

**Comment by MichaelDeBoey** [comment]
Closed by #3860

**Comment by Swapnull** [comment]
> > > @lswest @kokotom yeah i just tried the latest experimental and all is fine with that library :)
> > > [demo](https://pnpm-turborepo-vercel-remix-f65bs34pf-mcansh.vercel.app) [commit](https://github.com/mcansh/pnpm-turborepo-vercel-remix/commit/5847d74d0ffb447930674d03f104320fd9565d73)
> > 
> > 
> > Oh man terribly sorry. The package was: @syncfusion/ej2-react-gantt Glad you noticed it yourself :)
> > However I noticed that I didn't pin the version as mentioned by @lswest. After doing so I'm getting a different error:
> > `Warning: React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.`
> > It happened both with react 17.0.2 and 18.2.0
> 
> I have just tried `0.0.0-experimental-ea38b6458` after having these same problems and get this same error. ![image](https://user-images.githubusercontent.com/4456346/184081693-288d6b33-9caa-47c0-ad75-0750e59c37b1.png) My entry.server.tsx is exactly the same as the ones from the examples. I am on react 17.0.2 and macOS 12.0.1
> 
> The library I am having issues with is `@100mslive/react-sdk`

@chaance FYI updating to remix 1.7.0 now gives me this error and completely breaks my app in the same way I mentioned here (which I had working perfectly fine by using patch-package to remove some react imports from modules).

**Comment by chaance** [comment]
@Swapnull mind opening a new issue with a repro for this one?

**Comment by Swapnull** [comment]
@chaance Sure, for posterity, new issue opened #4080 with repro
